precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,446 @@
+	if !ok {
+		return nil, fmt.Errorf("invalid amount: %v", inputs[1])
+	}
+
+	if amount == nil {
+		return nil, errors.New("amount is required")
+	}
+
+	if amount.Sign() <= 0 {
+		return nil, errors.New("amount must be positive")
+	}
+
+	chainRaw, ok := inputs[2].(uint8)
+	if !ok {
+		return nil, fmt.Errorf("invalid chain: %v", inputs[2])
+	}
+
+	chain := TargetChain(chainRaw)
+
+	recipient, ok := inputs[3].([]byte)
+	if !ok {
+		return nil, fmt.Errorf("invalid recipient address: %v", inputs[3])
+	}
+
+	return &bridgeOutInputs{
+		Token:     token,
+		Amount:    amount,
+		Chain:     chain,
+		Recipient: recipient,
+	}, nil
+}
+
+func (m *BridgeOutMethod) validateRecipientForChain(chain TargetChain, recipient []byte) error {
+	if len(recipient) == 0 {
+		return fmt.Errorf("recipient can't be empty")
+	}
+
+	switch chain {
+	case TargetChainEthereum:
+		// here we just check the length, the zero address have been tested before already
+		if len(recipient) != 20 {
+			return fmt.Errorf("invalid recipient address format for Ethereum chain: %v", hex.EncodeToString(recipient))
+		}
+	case TargetChainBitcoin:
+		if keepbtc.GetScriptType(recipient) == keepbtc.NonStandardScript {
+			return fmt.Errorf("invalid recipient address format for Bitcoin: %v", hex.EncodeToString(recipient))
+		}
+	}
+
+	return nil
+}
+
+type TargetChain uint8
+
+const (
+	TargetChainEthereum = iota
+	TargetChainBitcoin
+)
+
+func (t TargetChain) Validate() (TargetChain, bool) {
+	if t == TargetChainEthereum || t == TargetChainBitcoin {
+		return t, true
+	}
+
+	return t, false
+}
+
+type bridgeOutInputs struct {
+	Token     common.Address
+	Amount    *big.Int
+	Chain     TargetChain
+	Recipient []byte
+}
+
+// AssetsUnlockedEventName is the name of the AssetsUnlocked event. It matches the name
+// of the event in the contract ABI.
+const AssetsUnlockedEventName = "AssetsUnlocked"
+
+// AssetsUnlockedEvent is the implementation of the AssetsUnlocked event that contains
+// the following arguments:
+// - unlockSequenceNumber (indexed): the sequenceNumber of this AssetsUnlocked
+// - recipient (indexed): the address to which the tokens are transferred
+// - token (indexed): the token being bridged out.
+// - sender (non-indexed): the address from which the tokens are bridged out,
+// - amount (non-indexed): the amount of tokens transferred
+// - chain (non-indexed): the destination chain
+type AssetsUnlockedEvent struct {
+	unlockSequenceNumber *big.Int
+	recipient            []byte
+	token                common.Address
+	sender               common.Address
+	amount               *big.Int
+	chain                uint8
+}
+
+func NewAssetsUnlockedEvent(
+	from, token common.Address,
+	recipient []byte,
+	chain uint8,
+	sequenceNumber *big.Int,
+	amount *big.Int,
+) *AssetsUnlockedEvent {
+	return &AssetsUnlockedEvent{
+		sender:               from,
+		recipient:            recipient,
+		token:                token,
+		unlockSequenceNumber: sequenceNumber,
```

Comment:
Suggestion to consider: passing arguments to `NewAssetsUnlockedEvent` and filling the fields in  `AssetsUnlockedEvent` could be done in the same order as they are declared in  
```
type AssetsUnlockedEvent struct { ... }
```
i.e.
```
	unlockSequenceNumber *big.Int
	recipient            []byte
	token                common.Address
	sender               common.Address
	amount               *big.Int
	chain                uint8
```

Reply:
+1 on this

Reply:
done in: 6f6f0faf